### PR TITLE
Whitelisttransactions

### PR DESCRIPTION
**Pull request by lawlawlaw**

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2780,8 +2796,18 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
             if(tx.vout[0].nAsset.GetAsset() == freezelistAsset) UpdateFreezeList(tx,view);
         }
         if(fEnableBurnlistCheck) {
-            if(tx.vout[0].nAsset.GetAsset() == burnlistAsset && fEnableBurnlistCheck) UpdateBurnList(tx,view);
+            if(tx.vout[0].nAsset.GetAsset() == burnlistAsset) UpdateBurnList(tx,view);
         }
+        //Non-whitelisting nodes need the kyc pub key whitelist as well
```
Comment (nkostoulas) [inline src/validation.cpp]: Why is this needed by all nodes?

File: Dockerfile [hunk 2]
```diff
@@ -8,7 +8,7 @@ RUN set -ex \
     && ./autogen.sh \
     && ./configure \
     && make clean \
-    && make  -j$(nproc)\
+    && make \
```
Comment (nkostoulas) [inline Dockerfile]: Could we leave this as before? It should allow faster builds.

**Comment by nkostoulas** [comment]
``` travis
$ if [ "$CHECK_DOC" = 1 ]; then contrib/devtools/check-doc.py; fi
Args used        : 204
Args documented  : 220
Args undocumented: 2
set(['-registeraddresssize', '-registeraddress'])
Args unknown     : 18
set(['-blockminsize', '-mainchainrpcpassword', '-fuzzmessagestest', '-help', '-version', '-zmqpubrawblock', '-relaypriority', '-h', '-kycmasterpubkey', '-mainchainrpcport', '-rpcconnect', '-zmqpubhashtx', '-zmqpubhashblock', '-zmqpubrawtx', '-mainchainrpchost', '-socks', '-mainchainrpcuser', '-?'])
The command "if [ "$CHECK_DOC" = 1 ]; then contrib/devtools/check-doc.py; fi" failed and exited with 2 during .
```

File: src/init.cpp [hunk 3]
```diff
@@ -527,6 +528,7 @@ std::string HelpMessage(HelpMessageMode mode)
         strUsage += HelpMessageOpt("-freezelistcoinsdestination", strprintf(_("The destination of the tokens for controlling the freezelist (default: %d)"), 0));
         strUsage += HelpMessageOpt("-burnlistcoinsdestination", strprintf(_("The destination of the tokens for controlling the burnlist. (default: %d)"), 0));
         strUsage += HelpMessageOpt("-whitelistcoinsdestination", strprintf(_("The destination of the tokens for controlling the whitelist. (default: %d)"), 0));
+        strUsage += HelpMessageOpt("-kycmasterpubkey", strprintf(_("The master public key of the wallet controlling the KYC public keys. (default: %d)"), 0));
```
[the remark this answers is not in the document]
Comment (lawlawlaw) [reply src/init.cpp]: Thanks for the feedback. This is not used.  I will remove it.

File: src/policy/policy.cpp [hunk 4]
```diff
@@ -120,8 +123,25 @@ bool IsPolicy(const CTransaction& tx)
...
+
+bool IsPolicy(const CAsset& asset){
+  if(asset == freezelistAsset || 
+     asset == burnlistAsset || 
+     asset == whitelistAsset) 
+    return true;
+  return false;
+}
+
+bool IsRegisterAddress(const CTransaction& tx){
```
[the remark this answers is not in the document]
Comment (lawlawlaw) [reply src/policy/policy.cpp]: That's no longer used - I will remove it.

File: src/init.cpp [hunk 5]
```diff
@@ -1090,12 +1095,17 @@ bool AppInitParameterInteraction()
 
     //address whitelisting
     fRequireWhitelistCheck = GetBoolArg("-pkhwhitelist", DEFAULT_WHITELIST_CHECK);
+    fScanWhitelist      = GetBoolArg("-pkhwhitelist-scan", DEFAULT_SCAN_WHITELIST);
```
[the remark this answers is not in the document]
Comment (lawlawlaw) [reply src/init.cpp]: Transactions being added to the mempool are only checked for whitelisted outputs if fRequireWhitelistCheck.
fScanWhitelist enables keeping a list of which of the wallet's addresses are whitelisted. It also enables updates to the list of unassigned KYC public keys which are published to the blockchain by the signing nodes (see CWhiteList::_kycUnassignedQueue). One of these keys is required to encrypt the onboarding KYC file (see dumpkycfile RPC).

File: src/validation.cpp (hunk 1 above)
Comment (lawlawlaw) [reply src/validation.cpp]: It's not strictly needed by all nodes but is needed for the reasons outlined in the above comment regarding fScanWhitelist.

File: Dockerfile (hunk 2 above)
Comment (lawlawlaw) [reply Dockerfile]: Ok

File: src/validation.cpp (hunk 1 above)
Comment (nkostoulas) [reply src/validation.cpp]: Ahhh cool

**Comment by lawlawlaw** [comment]
This would be needed if wanted to enable peer-peer connection rather than
just peer-server. The peers don't have full knowledge of the whitelist so
would block all transactions other than their own if fRequireWhitelist.

**Comment by nkostoulas** [comment]
I think you need the WALLET_ENABLED check in policy/whitelist.h as well to fix the last issue with travis.
